django/db/models/sql/where.py

```diff
@@ -231,6 +231,10 @@ class NothingNode:
     def as_sql(self, compiler=None, connection=None):
         raise EmptyResultSet
 
+    @staticmethod
+    def get_source_expressions():
+        return []
```

Comment:
I'd add checking `get_source_expressions` to the `_gen_cols()` instead, because other nodes can also be affected e.g. `ExtraWhere`:
```python
qs = Party.objects.all().annotate(
    foo=models.Count('id'),
    bar=models.Exists(Party.objects.extra(where=['1=0'])),
)
```
Maybe:
```diff
diff --git a/django/db/models/sql/query.py b/django/db/models/sql/query.py
index 1258eb61a9..39a735b50e 100644
--- a/django/db/models/sql/query.py
+++ b/django/db/models/sql/query.py
@@ -1692,7 +1692,7 @@ class Query(BaseExpression):
                 yield expr
             elif include_external and callable(getattr(expr, 'get_external_cols', None)):
                 yield from expr.get_external_cols()
-            else:
+            elif hasattr(expr, 'get_source_expressions'):
                 yield from cls._gen_cols(
                     expr.get_source_expressions(),
                     include_external=include_external,
```